### DISCUSSION TITLE
[Deployment] Deploying Beacon Proxy Upgradeable Simple ERC20 Contract

### DISCUSSION THREAD

**Discussion by vascovitche** [zkSync-Community-Hub/zksync-developers#221] [Deployments] [answered]
### Environment

Testnet

### zkSolc Version

1.3.17

### zksync-web3 Version

zksync2-js ^0.2.2

### Hardhat.config.ts

``` javascript
import { HardhatUserConfig } from "hardhat/config";

import "@matterlabs/hardhat-zksync-solc";
import "@matterlabs/hardhat-zksync-upgradable";

const config: HardhatUserConfig = {
  defaultNetwork: "zkSyncTestnet",
  networks: {
    zkSyncTestnet: {
      zksync: true,
      url: "https://testnet.era.zksync.dev",
      ethNetwork: "goerli",
      
    },
    zkSyncMainnet: {
      url: "https://mainnet.era.zksync.io",
      ethNetwork: "mainnet",
      zksync: true,
      
    },
    dockerizedNode: {
      url: "http://localhost:3050",
      ethNetwork: "http://localhost:8545",
      zksync: true,
    },
    inMemoryNode: {
      url: "http://127.0.0.1:8011",
      ethNetwork: "",
      zksync: true,
    },
    hardhat: {
      zksync: true,
    },
  },
  zksolc: {
    version: "1.3.17",
    settings: {
      // find all available options in the official documentation
      // https://era.zksync.io/docs/tools/hardhat/hardhat-zksync-solc.html#configuration
    },
  },
  solidity: {
    version: "0.8.20",
  },
};

export default config;
```

### Deployment Script (WITHOUT PRIVATE KEY)

``` javascript
import {Provider, Wallet,} from "zksync2-js";
import {Deployer} from "@matterlabs/hardhat-zksync-deploy";
import * as hre from "hardhat";

require('dotenv').config();

async function main() {
    const rpcUrl = hre.network.config.url;
    const provider = new Provider(rpcUrl);
    console.log(`RPC URL: ${rpcUrl}`);

    const zkWallet = new Wallet(process.env.WALLET_PRIVATE_KEY!);
    console.log(`Wallet address: ${zkWallet.address}`);

    const deployer = new Deployer(hre, zkWallet);

    const contractName = "Gus";
    const contract = await deployer.loadArtifact(contractName);

    const beacon = await hre.zkUpgrades.deployBeacon(deployer.zkWallet, contract);

    await beacon.deployed();

    console.log("Beacon deployed to:", beacon.address);

    const gus = await hre.zkUpgrades.deployBeaconProxy(deployer.zkWallet, beacon, contract, [0]);
    await gus.deployed();
    console.log(contractName + " beacon proxy deployed to: ", gus.address);

}

// We recommend this pattern to be able to use async/await everywhere
// and properly handle errors.
main().catch((error) => {
    console.error(error);
    process.exitCode = 1;
});
```

### Contract Code

```solidity
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.20;

import "@openzeppelin/contracts-upgradeable/token/ERC20/ERC20Upgradeable.sol";
import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";

contract Gus is Initializable, ERC20Upgradeable, OwnableUpgradeable {

    function initialize() initializer public {
        __ERC20_init("Gussiiii", "GUS");
        __Ownable_init();

        _mint(msg.sender, 200000 * 10 ** decimals());
    }

    function decimals() public pure override returns (uint8) {
        return 2;
    }
    
    // another functions

  }
```

### Does this work on other EVMs? (If yes, please list at least 1 of them)

Polygon, ZkSync (without proxy)

### Description of What Your Contract Does

Simple ERC20 Token. Burn, Mint.

### Additional Details

I want to deploy a simple ERC20 token contract using the proxy pattern as recommended documentation. I use **hardhat-zksync-upgradeable plugin (1.0.0)** and beacon proxy. Also, I use the simple example of deploying script from docks https://era.zksync.io/docs/tools/hardhat/hardhat-zksync-upgradable.html#full-code-for-deploy-beacon

In an example from docks using **zksync-web3**, but last version of **hardhat-zksync-deploy** working with **zksync2-js**.
So I have a trouble during the run script on this line:

```javascript
const beacon = await hre.zkUpgrades.deployBeacon(deployer.zkWallet, contract);
```

**Error:**
```error
Error: incorrect number of arguments to constructor
    at ContractFactory.getDeployTransaction (/home/ironp/dapp/proxy-step-by-step/node_modules/ethers/src.ts/contract/factory.ts:87:19)
    at ContractFactory.getDeployTransaction (/home/ironp/dapp/proxy-step-by-step/node_modules/zksync2-js/build/src/contract.js:55:39)
    at ContractFactory.deploy (/home/ironp/dapp/proxy-step-by-step/node_modules/ethers/src.ts/contract/factory.ts:105:31)
    at ContractFactory.deploy (/home/ironp/dapp/proxy-step-by-step/node_modules/zksync2-js/build/src/contract.js:125:38)
    at deploy (/home/ironp/dapp/proxy-step-by-step/node_modules/@matterlabs/hardhat-zksync-upgradable/src/proxy-deployment/deploy.ts:13:44)
    at Proxy.deployBeacon (/home/ironp/dapp/proxy-step-by-step/node_modules/@matterlabs/hardhat-zksync-upgradable/src/proxy-deployment/deploy-beacon.ts:46:88)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async main (/home/ironp/dapp/proxy-step-by-step/scripts/deploy.ts:20:20)
```

**Here is my package list:**

```json
  "devDependencies": {
    "@matterlabs/hardhat-zksync-deploy": "^1.0.0",
    "@matterlabs/hardhat-zksync-upgradable": "^1.0.0",
    "@openzeppelin/contracts": "^5.0.0",
    "@openzeppelin/contracts-upgradeable": "4.9.2",
    "dotenv": "^16.0.3",
    "ethers": "^6.9.0",
    "hardhat": "^2.19.1",
    "ts-node": "^10.9.1",
    "typescript": "^4.9.5",
    "zksync2-js": "^0.2.2"
  }
```

How can I fix it? Are somewhere examples with new versions of packages? Or what versions I must use for successful deploy by example in docks?

Thanks for answers

**Comment by vasyl-ivanchuk** [comment, marked as answer]
Hey @vascovitche 👋 ,

Thanks for reaching out. In the example you shared the are a couple of issues with the dependency versions and the deployment script. To make it work, I changed the dependencies to the following:
```JSON
  "devDependencies": {
    "@matterlabs/hardhat-zksync-deploy": "^1.0.0",
    "@matterlabs/hardhat-zksync-upgradable": "^1.0.0",
    "@openzeppelin/contracts": "^4.9.2",
    "@openzeppelin/contracts-upgradeable": "^4.9.2",
    "dotenv": "^16.0.3",
    "ethers": "^6.9.0",
    "hardhat": "^2.19.1",
    "ts-node": "^10.9.1",
    "typescript": "^4.9.5",
    "zksync-ethers": "^6.1.0"
  }
```

and modified the deployment script based on our `hardhat-zksync-upgradable` tutorial [here](https://docs.zksync.io/build/tooling/hardhat/hardhat-zksync-upgradable.html#full-code-for-deploy-beacon):
```JavaScript
import { Wallet } from "zksync-ethers";
import { Deployer } from "@matterlabs/hardhat-zksync-deploy";
import * as hre from "hardhat";

require('dotenv').config();

async function main() {
    const zkWallet = new Wallet(process.env.WALLET_PRIVATE_KEY!);
    console.log(`Wallet address: ${zkWallet.address}`);

    const deployer = new Deployer(hre, zkWallet);

    const contractName = "Box";
    const contract = await deployer.loadArtifact(contractName);

    const beacon = await hre.zkUpgrades.deployBeacon(deployer.zkWallet, contract);

    await beacon.waitForDeployment();

    const beaconAddress = await beacon.getAddress();

    const gus = await hre.zkUpgrades.deployBeaconProxy(deployer.zkWallet, beaconAddress, contract);
    await gus.waitForDeployment();
}

main().catch((error) => {
    console.error(error);
    process.exitCode = 1;
});
```

Feel free to check the full example [here](https://github.com/vasyl-ivanchuk/zksync-beacon-proxy-deployment-example).

**Comment by EagleGSII** [comment]
I would like to fix the issue with the following sentence: "I want to deploy a simple ERC20 token contract using the proxy pattern as recommended documentation."

-
To fix it, "I use hardhat-zksync-upgradeable plugin (1.0.0) and beacon proxy. Also, I use the simple example of deploying script" when I speak my mind out! I make a fucking huge effort to be clear whilst being honest (to my knowledge) at all times!

**Comment by candacekrank** [comment]
It seems that the error is due to an incorrect number of arguments passed to the constructor when deploying the contract beacon. Make sure that you are passing the correct arguments to the deployBeacon function. Review the documentation and examples provided by @matterlabs/hardhat-zksync-upgradable to ensure that you are using the correct syntax and parameters for deploying the beacon proxy contract. Additionally, consider checking for any updates or changes in the package versions that may affect the deployment process.

**Comment by maggielyon** [comment]
It seems like there might be an issue with the contract deployment script when using the `@matterlabs/hardhat-zksync-upgradable` plugin with `zksync2-js` library. Ensure that the contract deployment parameters are correctly passed to the `deployBeacon` function. Additionally, consider checking for any updates or changes in the plugin documentation or examples that might address this issue.

**Comment by educatedb** [comment]
It seems there's a discrepancy in the number of constructor arguments passed when deploying the beacon proxy. Ensure that the contract's constructor parameters are correctly provided in the deployment script.

**Comment by mancinottimoe** [comment]
It seems like there is an issue with the number of arguments passed to the constructor during deployment. Double-check the contract initialization and ensure that the correct number of arguments is being provided. Additionally, verify that the contract and deployment scripts are compatible with the versions of the libraries being used.
